# Exploring Ebay Car Sales Data

We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. 

The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which which year the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which which year the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.  

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv',encoding='Latin-1')

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [3]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

Or dataset contains 20 columns, most of which stored as strings. There are few columns with null values, but no columns have more than ~20% null values. There are some columns that contain dates as strings. We'll start by cleaning the column names to make data easier to work with

# Clean Columns

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll make few changes here:
* Change the columns from camelcase to snakecase.
* Change few wordings to describe the columns more clearly.

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year    gearbox  power_ps  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   odometer  registration_month fuel_type       brand unrepaired_damage  \
0    150000                   0    benzin  volkswagen               NaN   
1    125000                   5    diesel        audi                ja   
2    125000                   8    diesel        jeep               NaN   
3    150000                   6    benzin  volkswagen              nein   
4     90000                   7    diesel       skoda              nein   

            ad_created  num_photos  postal_code            last_seen  
0  2016-03-24 00:00:00           0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00           0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00           0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00           0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00           0        60437  2016-04-06 10:17:21

# Initial Data Exploration and Cleaning

We'll start by exploring the data to find obvious areas where we can clean the data.

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model       odometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count             299468               371528    371528.0  371528.00000   
unique                 2                  114         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq              263182                14450         NaN           NaN   
mean                 NaN                  NaN         0.0   50820.66764   
std                  NaN                  NaN         0.0   25799.08247   
min                  NaN                  NaN         0.0    1067.00000   
25%                  NaN                  NaN         0.0   30459.00000   
50%                  NaN                  NaN         0.0   49610.00000   
75%                  NaN                  NaN         0.0   71546.00000   
max                  NaN                  NaN         0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Initial observations:  
- The column seller and offer_type where nearly all the values are the same.  
- The num_photos column looks odd, we need to investigate further.  

In [7]:
autos['seller'].value_counts()

privat        371525
gewerblich         3
Name: seller, dtype: int64

In [8]:
autos['offer_type'].value_counts()

Angebot    371516
Gesuch         12
Name: offer_type, dtype: int64

In [9]:
autos['num_photos'].value_counts()

0    371528
Name: num_photos, dtype: int64

It looks like these columns dont have values. We'll drop these columns. 

In [10]:
autos.drop(['seller','offer_type','num_photos'], axis=1, inplace=True)

In [11]:
autos.head()

date_crawled                            name  price ab_test  \
0  2016-03-24 11:52:17                      Golf_3_1.6    480    test   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  18300    test   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"   9800    test   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER   1500    test   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic   3600    test   

  vehicle_type  registration_year    gearbox  power_ps  model  odometer  \
0          NaN               1993    manuell         0   golf    150000   
1        coupe               2011    manuell       190    NaN    125000   
2          suv               2004  automatik       163  grand    125000   
3   kleinwagen               2001    manuell        75   golf    150000   
4   kleinwagen               2008    manuell        69  fabia     90000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   
1                   5    diesel        audi                ja   
2                   8    diesel        jeep               NaN   
3                   6    benzin  volkswagen              nein   
4                   7    diesel       skoda              nein   

            ad_created  postal_code            last_seen  
0  2016-03-24 00:00:00        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00        60437  2016-04-06 10:17:21

In [12]:
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

In [13]:
autos.columns

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

# Exploring Odometer and Price

In [14]:
autos['odometer_km'].unique().shape

(13,)

In [15]:
autos['odometer_km'].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

In [16]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price, dtype: int64

In [17]:
autos['price'].value_counts().sort_index(ascending=True).head(20)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price, dtype: int64

There are a number of listings about 10.7k at $0. Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [18]:
autos[autos['price'].between(1,350000)]

date_crawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0         480     test          NaN               1993    manuell         0   
1       18300     test        coupe               2011    manuell       190   
2        9800     test          suv               2004  automatik       163   
3        1500     test   kleinwagen               2001    manuell        75   
4        3600     test   kleinwagen               2008    manuell        69   
...       ...      ...          ...                ...        ...       ...   
371523   2200     test          NaN               2005        NaN         0   
371524   1199     test       cabrio               2000  automatik       101   
371525   9200     test          bus               1996    manuell       102   
371526   3400     test        kombi               2002    manuell       100   
371527  28990  control    limousine               2013    manuell       320   

              model  odometer_km  registration_month fuel_type  \
0              golf       150000                   0    benzin   
1               NaN       125000                   5    diesel   
2             grand       125000                   8    diesel   
3              golf       150000                   6    benzin   
4             fabia        90000                   7    diesel   
...             ...          ...                 ...       ...   
371523          NaN        20000                   1       NaN   
371524       fortwo       125000                   3    benzin   
371525  transporter       150000                   3    diesel   
371526         golf       150000                   6    diesel   
371527      m_reihe        50000                   8    benzin   

                 brand unrepaired_damage           ad_created  postal_code  \
0           volkswagen               NaN  2016-03-24 00:00:00        70435   
1                 audi                ja  2016-03-24 00:00:00        66954   
2                 jeep               NaN  2016-03-14 00:00:00        90480   
3           volkswagen              nein  2016-03-17 00:00:00        91074   
4                skoda              nein  2016-03-31 00:00:00        60437   
...                ...               ...                  ...          ...   
371523  sonstige_autos               NaN  2016-03-14 00:00:00        39576   
371524           smart              nein  2016-03-05 00:00:00        26135   
371525      volkswagen              nein  2016-03-19 00:00:00        87439   
371526      volkswagen               NaN  2016-03-20 00:00:00        40764   
371527             bmw              nein  2016-03-07 00:00:00        73326   

                  last_seen  
0       2016-04-07 03:16:57  
1       2016-04-07 01:46:50  
2       2016-04-05 12:47:46  
3       2016-03-17 17:40:17  
4       2016-04-06 10:17:21  
...                     ...  
371523  2016-04-06 00:46:52  
371524  2016-03-11 18:17:12  
371525  2016-04-07 07:15:26  
371526  2016-03-24 12:45:21  
371527  2016-03-22 03:17:10  

[360635 rows x 17 columns]

# Exploring the date columns

There are a number of columns with date information:

* date_crawled
* registration_month
* registration_year
* ad_created
* last_seen

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.

In [19]:
autos[['date_crawled','ad_created','last_seen']][:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

In [20]:
autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_values()

2016-04-07    0.001620
2016-04-06    0.003163
2016-04-05    0.012823
2016-03-18    0.013127
2016-03-06    0.014462
2016-03-13    0.015735
2016-03-05    0.025551
2016-03-24    0.029912
2016-03-16    0.030148
2016-03-27    0.030275
2016-03-17    0.031661
2016-03-31    0.031885
2016-03-26    0.031965
2016-03-23    0.031971
2016-03-22    0.032469
2016-03-10    0.032563
2016-03-11    0.032727
2016-03-25    0.032934
2016-03-15    0.033451
2016-03-08    0.033454
2016-03-30    0.033529
2016-04-01    0.034116
2016-03-29    0.034170
2016-03-09    0.034210
2016-04-02    0.035079
2016-03-28    0.035112
2016-03-19    0.035292
2016-03-07    0.035690
2016-03-21    0.035739
2016-03-12    0.036194
2016-03-14    0.036275
2016-03-20    0.036353
2016-04-04    0.037612
2016-04-03    0.038735
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [21]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001292
2016-03-06    0.004134
2016-03-07    0.005262
2016-03-08    0.008056
2016-03-09    0.009994
2016-03-10    0.011563
2016-03-11    0.013046
2016-03-12    0.023401
2016-03-13    0.008489
2016-03-14    0.012301
2016-03-15    0.016411
2016-03-16    0.016419
2016-03-17    0.028760
2016-03-18    0.006931
2016-03-19    0.016314
2016-03-20    0.019915
2016-03-21    0.020136
2016-03-22    0.020607
2016-03-23    0.018149
2016-03-24    0.019237
2016-03-25    0.019097
2016-03-26    0.016160
2016-03-27    0.016909
2016-03-28    0.022273
2016-03-29    0.023312
2016-03-30    0.023856
2016-03-31    0.024238
2016-04-01    0.024022
2016-04-02    0.025016
2016-04-03    0.025366
2016-04-04    0.025654
2016-04-05    0.126206
2016-04-06    0.217830
2016-04-07    0.129648
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

In [22]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.038893
2016-04-04    0.037741
2016-04-05    0.011655
2016-04-06    0.003155
2016-04-07    0.001558
Name: ad_created, Length: 114, dtype: float64

There is a large variety of ad created dates. Oldest around 2 years ago.

In [23]:
autos['registration_year'].describe()

count    371528.000000
mean       2004.577997
std          92.866598
min        1000.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd calues. The min value is 1000 long before the cars were invented and the maximum is 9999, many years into future.

# Dealing with Incorrect Registration Year Data

The first car around decades of 1900s and above 2016 is inaccurate as car cant be registered before listing

One option is to remove the listings with these values. Let's determine what percentage our data has invalid values in this column.

In [24]:
(~autos['registration_year'].between(1900, 2016)).sum()/autos.shape[0]

0.03969552765874981

Given that this is less than 4% of our data, we will remove these rows.

In [25]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos['registration_year'].value_counts(normalize=True).head(10)

2000    0.068813
1999    0.063812
2005    0.062548
2006    0.056702
2001    0.056668
2003    0.055701
2004    0.055345
2002    0.053784
1998    0.050314
2007    0.049535
Name: registration_year, dtype: float64

# Exploring Price by Brand

In [26]:
autos['brand'].value_counts(normalize=True).sort_values(ascending=False).head(20)

volkswagen        0.212391
bmw               0.109695
opel              0.107069
mercedes_benz     0.095955
audi              0.089352
ford              0.068849
renault           0.047559
peugeot           0.029859
fiat              0.025758
seat              0.018631
skoda             0.015407
mazda             0.015348
smart             0.014104
citroen           0.013874
nissan            0.013569
toyota            0.012747
sonstige_autos    0.010648
hyundai           0.009830
mini              0.009210
volvo             0.009129
Name: brand, dtype: float64

German manufacturers represent the top five brands, almost 50% of overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of next two brands combined.

There are lots of brands that dont have significant percentage of listings, we can limit our analysis to brands representing more than 5% of total listings.

In [27]:
brand_counts = autos['brand'].value_counts(normalize=True)
common_brands = brand_counts[brand_counts>0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [28]:
brand_mean_price = {}

for brand in common_brands:
    brand_cars = autos[autos['brand']==brand]
    mean_price = brand_cars['price'].mean()
    brand_mean_price[brand] = int(mean_price)
brand_mean_price

{'volkswagen': 13643,
 'bmw': 14798,
 'opel': 3248,
 'mercedes_benz': 17614,
 'audi': 16218,
 'ford': 8702}

Of the top 5 brands, there is a distinct price gap:
* Audi, BMW & Mercedes Benz are more expensive
* Ford and Opel are less expensive
* Volkswagen is in between- this may explain its popularity, best for both world's option

In [29]:
bmp_series = pd.Series(brand_mean_price)
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
volkswagen          13643
bmw                 14798
opel                 3248
mercedes_benz       17614
audi                16218
ford                 8702